### PR TITLE
okhttp: using pending stream to queue writings before the connection is connected.

### LINKED ISSUE #759: okhttp: using pending stream to queue writings before the connection is connected.
Currently we use the serializing executor to implicitly queue writings before the connection is connected.

As we've added an obvious queue inside pending stream, using it to queue writings during connecting makes more sense.

### PR DESCRIPTION
**Pull request by madongfly**
Fixes #759

### CODE REVIEW

File: okhttp/src/main/java/io/grpc/transport/okhttp/AsyncFrameWriter.java [hunk 1]
```diff
@@ -184,6 +189,10 @@ public void ping(final boolean ack, final int payload1, final int payload2) {
     executor.execute(new WriteRunnable() {
       @Override
       public void doRun() throws IOException {
+        if (frameWriter == null) {
+          // In case the upper layer calls ping when the connection hasn't been connected yet.
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/AsyncFrameWriter.java]: This orphans the ping. It seems handling it like any other write would be fine, i.e., throw an exception.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java [hunk 2]
```diff
@@ -263,18 +264,25 @@ public OkHttpClientStream newStream(MethodDescriptor<?, ?> method,
...
-    stream.start(nextStreamId);
+    int streamId = nextStreamId;
+    if (nextStreamId >= Integer.MAX_VALUE - 2) {
+      // Make sure nextStreamId is greater than all used ids, so that mayHaveCreatedStream()
+      // performs correctly.
+      nextStreamId = Integer.MAX_VALUE;
+    } else {
+      nextStreamId += 2;
+    }
+    streams.put(streamId, stream);
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: This seems broken. Doesn't this actually create new streams after nextStreamId == MAX_VALUE? That means that we will then re-use MAX_VALUE for the streamId, which is _very bad_.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java [hunk 3]
```diff
@@ -305,73 +313,70 @@ void removePendingStream(OkHttpClientStream pendingStream) {
...
+            maxConcurrentStreams = Integer.MAX_VALUE;
+          }
+          startPendingStreams();
+          connectedFuture.set(null);
           return;
         }
+
         BufferedSource source;
         BufferedSink sink;
-        Socket sock;
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: The socket field isn't used any more. Delete it and just use a local variable here.

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java [hunk 4]
```diff
@@ -263,18 +264,25 @@ public OkHttpClientStream newStream(MethodDescriptor<?, ?> method,
   @GuardedBy("lock")
   private void startStream(OkHttpClientStream stream) {
     Preconditions.checkState(stream.id() == null, "StreamId already assigned");
-    streams.put(nextStreamId, stream);
-    stream.start(nextStreamId);
+    int streamId = nextStreamId;
+    if (nextStreamId >= Integer.MAX_VALUE - 2) {
+      // Make sure nextStreamId is greater than all used ids, so that mayHaveCreatedStream()
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java]: Add/enhance two tests:

1. Send data and half-close on a pending stream. You may optionally (in addition) remove the `mayHaveCreatedStream()` check in `OutboundFlowController`, but it is fine either way.
2. Add a test for receiving headers/data/trailers on stream with id MAX_VALUE-2

**Comment by ejona86** [comment]
Discussed offline. I understand a lot more of what is going on. We are going to postpone the `serializingExecutor` changes and make two small PRs, one with the nextStreamId fix for pending streams and one for setting max concurrent streams = 0 before connecting.

**Comment by madongfly** [comment]
Per our offline discussion, I split ddea7435c9952e4ee8ecaf8f52bda6733a7bc824 and b1e2aaebc066b10b00711b51b8cab0b09151544f from this PR.

But we are still using the `serializingExecutor` for connecting, since it's dangerous to change.

I'll keep this change in branch `madongfly:async` for reference if we decide to remove the usage of `serializingExecutor` in the future.

**Pull request closed without merging**